### DISCUSSION TITLE
Wrong density matrix

### DISCUSSION THREAD

**Discussion by kaanyurtseven112358** [pyscf/pyscf#2658] [Q&A]
Hello, I am new to pySCF. I want to create density matrices from creation and annihilation operators. However, the code is not giving the meaningful results. 

import numpy
from pyscf import gto, scf, fci

M = gto.M(atom='H 0 0 0; H 0 0 0.740', basis='sto-3g')

m = dft.RKS(M).run()
fs = fci.FCI(M, m.mo_coeff)
e, c = fs.kernel()

norb = m.mo_energy.size
neleca = nelecb = M.nelectron // 2

dm1 = numpy.zeros((norb,norb))
for i in range(norb):
    for j in range(norb):
        tmp = fci.addons.des_a(c  , norb, (neleca  ,nelecb), j)
        tmp = fci.addons.cre_a(tmp, norb, (neleca-1,nelecb), i)
        dm1[i,j] += numpy.dot(tmp.flatten(), c.flatten())

        tmp = fci.addons.des_b(c  , norb, (neleca,nelecb  ), j)
        tmp = fci.addons.cre_b(tmp, norb, (neleca,nelecb-1), i)
        dm1[i,j] += numpy.dot(tmp.flatten(), c.flatten())

dm2aa = numpy.zeros((norb,norb))
dm2ab = numpy.zeros((norb,norb))
dm2bb = numpy.zeros((norb,norb))
for i in range(norb):
    for j in range(norb):
 
            tmp = fci.addons.des_a(c  , norb, (neleca  ,nelecb), j)
            tmp = fci.addons.cre_a(tmp, norb, (neleca-1,nelecb), i)
            dm2aa[i,j] += numpy.dot(tmp.flatten(), c.flatten())

            tmp = fci.addons.des_a(c  , norb, (neleca  ,nelecb  ), i)
            tmp = fci.addons.des_b(tmp, norb, (neleca-1,nelecb  ), j)
            tmp = fci.addons.cre_b(tmp, norb, (neleca-1,nelecb-1), j)
            tmp = fci.addons.cre_a(tmp, norb, (neleca-1,nelecb  ), i)
            dm2ab[i,j] += numpy.dot(tmp.flatten(), c.flatten())

            tmp = fci.addons.des_b(c  , norb, (neleca,nelecb  ), j)
            tmp = fci.addons.cre_b(tmp, norb, (neleca,nelecb-1), i)
            dm2bb[i,j] += numpy.dot(tmp.flatten(), c.flatten())

ref1 = fs.make_rdm1(c, norb, (neleca,nelecb))
ref2aa, ref2ab, ref2bb = fs.make_rdm12s(c, norb, (neleca,nelecb))[1]
ref1, ref2aa,ref2ab,ref2bb
print('Error in spin-free 1-PDM %g' % numpy.linalg.norm(ref1-dm1))
print('Error in 2-PDM aa %g' % numpy.linalg.norm(ref2aa.transpose(0,2,1,3)-dm2aa))
print('Error in 2-PDM ab %g' % numpy.linalg.norm(ref2ab.transpose(0,2,1,3)-dm2ab))
print('Error in 2-PDM bb %g' % numpy.linalg.norm(ref2bb.transpose(0,2,1,3)-dm2bb))

Errors are highly great for this simple exercise. Where am I making mistakes?
converged SCF energy = -1.12120611516684
Error in spin-free 1-PDM 0
Error in 2-PDM aa 1.97483
Error in 2-PDM ab 1.78147
Error in 2-PDM bb 1.97483

**Comment by sunqm** [comment]
`dm2`s are two-particle density matrix, a four-dimensional array.

**Comment by kaanyurtseven112358** [comment]
Thank you, but when we add two des_b or des_a the total number of electrons will be zero for the molecule. Then, this code won't work. Am I missing something?

**Comment by sunqm** [comment]
Functions like des_b, des_a support zero-number of electrons. You can keep applying these operators to the wave function.
